src/helpers/audio/worker.ts

```diff
@@ -121,25 +134,48 @@ function drawSpectrogramOntoDestinationCanvas(_generation: number): void {
   // eslint-disable-next-line @typescript-eslint/ban-ts-comment
   // @ts-ignore
   destinationSurface.commit && destinationSurface.commit();
+
+  // We only set hasBufferedImage to false after successfully drawing to the
+  // canvas so if we fail to draw (for any reason), it will automatically retry
+  // on the next update.
+  // Note that I have never seen this happen in practice, but I have included
+  // this logic as a defensive programming measure.
+  hasBufferedImage = false;
 }
 
-function setup(data: SharedBuffersWithCanvas): void {
+function setup(data: SharedBuffers): void {
   //console.log("worker:setup:", data);
   state = new WorkerState(data.state);
   sampleBuffer = new Float32Array(data.sampleBuffer);
-  destinationCanvas = data.canvas;
 
   // just use a default size - regenerate will resize it in a second
   spectrogramCanvas = new OffscreenCanvas(512, 512);
   spectrogramSurface = spectrogramCanvas.getContext("2d") as OffscreenCanvasRenderingContext2D;
 
-  destinationSurface = destinationCanvas.getContext("2d") as OffscreenCanvasRenderingContext2D;
-  destinationSurface.imageSmoothingEnabled = true;
-  destinationSurface.imageSmoothingQuality = "high";
+  // Because each pixel in the spectrogram is meaningful, we disable image
+  // smoothing so that the data is visible without any subtle modification.
+  spectrogramSurface.imageSmoothingEnabled = false;
 
   state.workerReady();
 }
 
+function setupDestinationCanvas(canvas: OffscreenCanvas): void {
+  destinationCanvas = canvas;
+
+  destinationSurface = destinationCanvas.getContext("2d", {
+    alpha: false,
+  }) as OffscreenCanvasRenderingContext2D;
+  destinationSurface.imageSmoothingEnabled = false;
```

Comment:
this is a secondary change in an optimization PR, it does not belong here.

You're right, each pixel is meaningful, but in our source buffer, not in our destination canvas. Because we have stretch and natural sizing we want to take advantage of image smoothing. Only on the original size, or in sizes that are exact scaled intervals of the original size do we probably want to disable this.

We did actually test this when we first set up the project.

We can revisit, in another PR, with before and after shots, for various sizes and sizing modes